core/src/main/java/com/graphhopper/routing/weighting/custom/CustomWeighting.java

```diff
@@ -0,0 +1,122 @@
+ * speed = reduce_to_max_speed(estimated_average_speed * multiply_all(speed_factor_map))
+ * weight = toSeconds(distance / speed) / multiply_all(priority_map) + distance * distance_influence;
+ * return weight
+ * </pre>
+ * Please note that the max_speed map is capped to the maximum allowed speed. Also values in the speed_factor and
+ * priority maps are normalized via the maximum value to 1 if one of the values is bigger than 1
+ * to avoid problems with the landmark algorithm, i.e. the edge weight is always increased so that the heuristic always
+ * underestimates the weight.
+ */
+public final class CustomWeighting extends AbstractWeighting {
+    public static final String NAME = "custom";
+
+    /**
+     * Converting to seconds is not necessary but makes adding other penalties easier (e.g. turn
+     * costs or traffic light costs etc)
+     */
+    private final static double SPEED_CONV = 3.6;
+    private final BooleanEncodedValue baseVehicleAccessEnc;
+    private final double maxSpeed;
+    private final double distanceInfluence;
+    private final double headingPenaltySeconds;
+    private final SpeedCustomConfig speedConfig;
+    private final PriorityCustomConfig priorityConfig;
+
+    public CustomWeighting(FlagEncoder baseFlagEncoder, EncodedValueLookup lookup,
+                           TurnCostProvider turnCostProvider, CustomModel customModel) {
+        super(baseFlagEncoder, turnCostProvider);
+        if (customModel == null)
+            throw new IllegalStateException("CustomModel cannot be null");
+
+        headingPenaltySeconds = customModel.getHeadingPenalty();
+        baseVehicleAccessEnc = baseFlagEncoder.getAccessEnc();
+        speedConfig = new SpeedCustomConfig(baseFlagEncoder.getMaxSpeed(), customModel, baseFlagEncoder.getAverageSpeedEnc(), lookup);
+        maxSpeed = speedConfig.getMaxSpeed() / SPEED_CONV;
+
+        priorityConfig = new PriorityCustomConfig(customModel, lookup);
+
+        // unit is "seconds per 1km"
+        distanceInfluence = customModel.getDistanceInfluence() / 1000;
+        if (distanceInfluence < 0)
+            throw new IllegalArgumentException("maximum distance_influence cannot be negative " + distanceInfluence);
+    }
+
+    @Override
+    public double getMinWeight(double distance) {
+        return distance / maxSpeed + distance * distanceInfluence;
+    }
+
+    @Override
+    public double calcEdgeWeight(EdgeIteratorState edgeState, boolean reverse) {
+        double distance = edgeState.getDistance();
+        double seconds = calcSeconds(distance, edgeState, reverse);
+        if (Double.isInfinite(seconds))
+            return Double.POSITIVE_INFINITY;
+        double distanceInfluence = distance * this.distanceInfluence;
+        if (Double.isInfinite(distanceInfluence))
+            return Double.POSITIVE_INFINITY;
+        return seconds / priorityConfig.calcPriority(edgeState, reverse) + distanceInfluence;
```

Comment:
I think its problematich to divide the time by priority while distanceInfluence is not affected by the priority. I think the formula should read:

```
weight = distance/speed + distance/priority + distance * distance_influence
```

To achieve 'preferring' certain roads (here motorway) we should *decrease* the priority of *all other* roads, which we should be able to realize using a match-all key that specifies the priority for all enum keys except the ones explicitly listed in the custom model, like: `priority: { road_class: {motorway: 100, "*": 50}}`. Note that with this formula the 'priority' needs to be interpreted as 'virtual speed' somehow, i.e. its value range should be of the same order as the speeds. If this is too hard for a user to understand we could still make the user enter some 'scaling' values like 0.8 and convert these values internally.

Reply:
This sounds good. Would you mind to see if you get the existing tests working with this?

> If this is too hard for a user to understand we could still make the user enter some 'scaling' values like 0.8 and convert these values internally

When we avoid "**prefer**" then values smaller than 1 are no longer possible and necessary. I.e. the only thing we have to tell the use is the acceptable range, which can be `[1,Infinity]` or `[100,Infinity]` or whatever. The main thing is that this is the same for all "maps".

Reply:
> Would you mind to see if you get the existing tests working with this?

Ok sure.

> When we avoid "prefer" then values smaller than 1 are no longer possible and necessary.  

Not sure if I understand. Why are values <1 no longer possible or necessary? For example in above formula priority 0 still means 'block this edge'.

>  I.e. the only thing we have to tell the use is the acceptable range

Yes the acceptable range for priority would be 0 to infinity. The thing the user has to understand is that the relative size of speed compared to priority and distance influence matters. So instead of being like 'I want this to be a little less prioritized, I set it to '90%'(=0.9)', one would have to give an absolute value of the cost (unless of course we somehow do this conversion internally). For example if the speed is 50, the per-meter cost is 1/50=0.02 and setting priority=3 would mean the per meter priority cost would be 1/3=0.33, i.e. 15x the speed costs and in this case the speed costs almost do not matter anymore. To still respect the speed the priority would have to be more in the range of 50, such that it does not totally dominate the speed costs. But I will see if I can come up with something handy.

> The main thing for me is that this is the same for all "factor maps" (speed_factor, priority).

What do you mean with 'factor maps'? speed_factor just specifies a factor that is multiplied with the default speed yes, but what would we be multiplying the priority with?

Reply:
Probably this message is display 10 times ... trying it again as I cannot come through:

> Not sure if I understand. Why are values <1 no longer possible or necessary? For example in above formula priority 0 still means 'block this edge'

Sorry, I was wrong. I think a range [0,1] should be sufficient. Because values bigger 1 were only introduced to **prefer** something. But with the match-all-key this is no longer necessary (?)

> Yes the acceptable range for priority would be 0 to infinity.

Would [0,1] be sufficient?

> What do you mean with 'factor maps'? speed_factor just specifies a factor that is multiplied with the default speed yes, but what would we be multiplying the priority with?

I just wanted to exclude the max_speed map where the unit is km/h.

The good thing is that we can currently specify 0.5 for priority or 0.5 for speed_factor and both would have the effect of doubling the weight. Would be nice if this "intuitive influence" property can be done for the new formula without the need for an unfamiliar unit.

Reply:
Yes this is the key question here. For speed its quite natural to use a factor, because there is something we can multiply the factor with (the speed), but for priority it is not that obvious.

Reply:
We could assume a default per-meter cost for every edge and multiply this with the priority factor (which by default would be 1). This would change the meaning of the distance_influence factor:

```
w = d/v + d*s + d*f  # d=distance, v=speed, s=stress_per_distance, f='old'_distance_influence
  = d/(v_base*speed_factor) + d/(x*priority) + d*distance_influence

speed_factor = v/v_base
priority = 1/(1+s*x)
distance_influence = f-1/x
x > 0
```

So we can:

* apply a factor to the base velocity (speed_factor)
* apply a factor to the base per-distance cost x (priority)
* the new distance_influence factor is reduce by 1/x, but it still has the same effect: the larger it is the less likely long routes will be (regardless of the road properties)

Then the only question is which value we use for x. Let's say its much smaller than v_base then a priority of 1 (which we want to make the default) means that the priority term has a lot of weight and the actual speed does no longer matter that much. If it is much larger than v_base its the other way around: Even a low priority (like 0.5) would mean that the second term is small compared to the first and the priority does not really change the routing behavior. So a good value should be some kind of 'average speed' of the vehicle, maybe like 30% of the maximum speed or something. Also note that even when setting distance_influence to 0 using, e.g. x=30 means the routing acts like with the old formula for p=1 and distance_influence=30, i.e. we cannot totally get rid of the distance influence unless we set all the priorities to infinity. But a value of x=30 would not hurt that much because the default distance_influence is 70 at the moment anyway. x should just not be too high (and not too low) :) 

We could also make x configurable but its probably too hard to understand.

Reply:
Oh and one thing I missed so far is that because of LM restrictions we cannot **decrease** weights, right? I.e. speed factor must always be `<1`. And this is the same for priority then. Preferring some road types then works by **reducing** the priority on all other roads (using the match-all key...)

Reply:
Btw its a little sad that we are not able to increase speeds (`speed_factor` > 1). It should be possible somehow by decreasing all speeds and then for time calculation to the backward conversion (?), but maybe its time to stop adding more possibilities for some time.

Reply:
>  It should be possible somehow by decreasing all speeds and then for time calculation to the backward conversion (?), but maybe its time to stop adding more possibilities for some time.

Yes and we can add this feature later without a problem.

> apply a factor to the base velocity (speed_factor)
> apply a factor to the base per-distance cost x (priority)

But is it then possible with "priority" (like with the current formula) to avoid motorways without changing the ETA?

Reply:
> But is it then possible with "priority" (like with the current formula) to avoid motorways without changing the ETA?

Yes of course. For the time calculation we (still) only take into account the speed and neither priority nor distance_influence.

Reply:
For custom weighting we do not use the vehicle priority encoded values (like `bike.priority`), right?

Reply:
For now not. Because those non-shared values could often be re-created using surface and bike_network encoded values. At least in theory. Maybe we need them later, but not now.